File: modules/engagement/android/src/main/java/expo/modules/engagement/EngagementSampleBuffer.kt
package expo.modules.engagement

class EngagementSampleBuffer {
  private val store = mutableMapOf<String, MutableList<EngagementSample>>()

  fun ensureKey(postId: String) {
    if (!store.containsKey(postId)) {
      store[postId] = mutableListOf()
    }
  }

  fun append(postId: String, sample: EngagementSample) {
    val list = store.getOrPut(postId) { mutableListOf() }
    list.add(sample)
  }

  fun appendToAll(postIds: Iterable<String>, sample: EngagementSample) {
    for (postId in postIds) {
      append(postId, sample)
    }
  }

  fun snapshot(): Map<String, MutableList<EngagementSample>> = store

  fun clearFor(postId: String) {
    store.remove(postId)
  }

  fun clearAll() {
    store.clear()
  }
}

File: modules/engagement/ios/EngagementSampleBuffer.swift
import Foundation

final class EngagementSampleBuffer {
  private var store: [String: [EngagementSample]] = [:]

  func ensureKey(_ postId: String) {
    if store[postId] == nil {
      store[postId] = []
    }
  }

  func append(_ postId: String, _ sample: EngagementSample) {
    if store[postId] == nil {
      store[postId] = []
    }
    store[postId]?.append(sample)
  }

  func appendToAll(_ postIds: Set<String>, _ sample: EngagementSample) {
    for postId in postIds {
      append(postId, sample)
    }
  }

  func snapshot() -> [String: [EngagementSample]] {
    return store
  }

  func clearFor(_ postId: String) {
    store.removeValue(forKey: postId)
  }

  func clearAll() {
    store.removeAll()
  }
}
